File: backend-node/src/controllers/dashboardController.ts
import type { Request, Response } from "express";
import { normalizeParam } from "../utils/params.js";
import { getAdminDashboardState, getTouristDashboard } from "../services/dashboardService.js";

export async function adminDashboard(_req: Request, res: Response) {
  const data = await getAdminDashboardState();
  res.json(data);
}

export async function touristDashboard(req: Request, res: Response) {
  const touristId = normalizeParam(req.params.touristId);
  if (!touristId) {
    return res.status(400).json({ message: "Invalid tourist ID." });
  }
  const data = await getTouristDashboard(touristId);
  if (!data) {
    return res.status(404).json({ message: "Tourist not found" });
  }
  return res.json(data);
}

File: backend-spring/src/main/java/com/safarsathi/controller/DashboardController.java
package com.YatraX.controller;

import com.YatraX.dto.AdminDashboardResponse;
import com.YatraX.dto.TouristDashboardResponse;
import com.YatraX.service.DashboardService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api")
@RequiredArgsConstructor
public class DashboardController {

    private final DashboardService dashboardService;

    /**
     * GET /api/admin/dashboard/state
     */
    @GetMapping("/admin/dashboard/state")
    public ResponseEntity<AdminDashboardResponse> adminDashboard() {
        return ResponseEntity.ok(dashboardService.getAdminDashboardState());
    }

    /**
     * GET /api/tourist/{touristId}/dashboard
     */
    @GetMapping("/tourist/{touristId}/dashboard")
    public ResponseEntity<?> touristDashboard(@PathVariable String touristId) {
        if (touristId == null || touristId.isBlank()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Invalid tourist ID."));
        }
        TouristDashboardResponse data = dashboardService.getTouristDashboard(touristId);
        if (data == null) {
            return ResponseEntity.status(404).body(Map.of("message", "Tourist not found"));
        }
        return ResponseEntity.ok(data);
    }
}
